File: todos-java/src/main/java/de/welcz/todosjava/Controller.java
package de.welcz.todosjava;

import org.bson.types.ObjectId;
import org.jetbrains.annotations.NotNull;
import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.Link;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.net.URI;
import java.text.MessageFormat;
import java.util.Objects;

@RestController
@RequestMapping("/api/v1/todos")
public class Controller {
  private final TodosRepository repository;

  public Controller(TodosRepository repository) {
    this.repository = repository;
  }

  @GetMapping("{id}")
  public Mono<EntityModel<Todo>> getTodo(@PathVariable ObjectId id) {
    return repository.findById(id)
                     .map(this::withHateoas)
                     .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NO_CONTENT)));
  }

  @PostMapping
  public Mono<ResponseEntity<EntityModel<Todo>>> createTodo(@RequestBody Todo toSave) {
    return repository.save(toSave)
                     .map(this::withHateoas)
                     .map(this::wrapInCreatedResponse);
  }

  @PutMapping("{id}")
  public Mono<EntityModel<Todo>> updateTodo(@PathVariable ObjectId id, @RequestBody Todo toSave) {
    return repository.findById(id)
                     .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NO_CONTENT)))
                     .then(Mono.fromCallable(() -> toSave.withId(id)))
                     .flatMap(repository::save)
                     .map(this::withHateoas);
  }

  @ResponseStatus(HttpStatus.NO_CONTENT)
  @DeleteMapping("{id}")
  public Mono<Void> deleteTodo(@PathVariable ObjectId id) {
    return repository.deleteById(id);
  }

  @GetMapping
  public Flux<EntityModel<Todo>> getAllTodos() {
    return repository.findAll()
                     .map(this::withHateoas);
  }

  private EntityModel<Todo> withHateoas(Todo todo) {
    return EntityModel.of(todo, Link.of(MessageFormat.format("/api/v1/todos/{0}", todo.getId()))
                                    .withSelfRel());
  }

  @NotNull
  private ResponseEntity<EntityModel<Todo>> wrapInCreatedResponse(EntityModel<Todo> todo) {
    return ResponseEntity.created(URI.create(MessageFormat.format("/api/v1/todos/{0}",
                                                                  Objects.requireNonNull(todo.getContent()).getId())))
                         .body(todo);
  }
}

File: todos-coroutini/src/main/kotlin/de/welcz/todoscoroutini/Controller.kt
package de.welcz.todoscoroutini

import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.reactive.awaitSingle
import org.bson.types.ObjectId
import org.springframework.hateoas.CollectionModel
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.reactive.WebFluxLinkBuilder.linkTo
import org.springframework.hateoas.server.reactive.WebFluxLinkBuilder.methodOn
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import java.net.URI

const val root = "/api/v1/todos"

@RestController
@RequestMapping(root)
class Controller(private val repository: TodoRepository) {
  @GetMapping("{id}")
  suspend fun getTodo(@PathVariable id: ObjectId) =
    repository.findById(id)?.withSelfLink()
      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)

  @PostMapping
  suspend fun createTodo(@RequestBody toSave: Todo) =
    repository.save(toSave).withSelfLink().wrappedInCreatedResponse()

  @PutMapping("{id}")
  suspend fun updateTodo(@PathVariable id: ObjectId, @RequestBody toModify: Todo) =
    repository.findById(id)
      ?.let { repository.save(toModify.copy(id = id)) }
      ?.withSelfLink()
      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)

  @ResponseStatus(HttpStatus.NO_CONTENT)
  @DeleteMapping("{id}")
  suspend fun deleteTodo(@PathVariable id: ObjectId) =
    repository.deleteById(id)

  @GetMapping
  suspend fun getAllTodos(): CollectionModel<EntityModel<Todo>> =
    repository
      .findAll()
      .map { it.withSelfLink() }
      .toList()
      .withSelfLink()

  private suspend fun <T> List<T>.withSelfLink(): CollectionModel<T> {
    val selfLink = linkTo(methodOn(Controller::class.java).getAllTodos())
      .withSelfRel()
      .toMono()
      .awaitSingle()
    return CollectionModel.of(this.toList(), selfLink)
  }

  private suspend fun Todo.withSelfLink(): EntityModel<Todo> {
    val selfLink = linkTo(methodOn(Controller::class.java).getTodo(id!!))
      .withSelfRel()
      .toMono()
      .awaitSingle()
    return EntityModel.of(this, selfLink)
  }

  private fun EntityModel<Todo>.wrappedInCreatedResponse() =
    ResponseEntity.created(URI("$root/${content!!.id}")).body(this)
}
